### PR TITLE
Make Math.Pow, Min and Max for doubles intrinsics

### PR DESCRIPTION
**Pull request by EgorBo**
Before this PR `Pow`, `Min` and `Max` worked only for `float`

```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static float Test1(float x) => MathF.Pow(x, 2);

[MethodImpl(MethodImplOptions.NoInlining)]
static double Test2(double x) => Math.Pow(x, 2);

// Min and Max require `--ffast-math` mode for optimal codegen

[MethodImpl(MethodImplOptions.NoInlining)]
static float Test3(float x, float y) => MathF.Min(x, y);

[MethodImpl(MethodImplOptions.NoInlining)]
static double Test4(double x, double y) => Math.Min(x, y);

[MethodImpl(MethodImplOptions.NoInlining)]
static float Test5(float x, float y) => Math.Max(x, y);

[MethodImpl(MethodImplOptions.NoInlining)]
static double Test6(double x, double y) => Math.Max(x, y);
```

Asm:
```asm

0000000000000000	vmulss	%xmm0, %xmm0, %xmm0
0000000000000004	retq

0000000000000000	vmulsd	%xmm0, %xmm0, %xmm0
0000000000000004	retq

0000000000000000	vminss	%xmm0, %xmm1, %xmm0
0000000000000004	retq

0000000000000000	vminsd	%xmm0, %xmm1, %xmm0
0000000000000004	retq

0000000000000000	vmaxss	%xmm0, %xmm1, %xmm0
0000000000000004	retq

0000000000000000	vmaxsd	%xmm0, %xmm1, %xmm0
0000000000000004	retq
```

LLVM IR: (Optimized, after InstCombine, etc)
```llvm
define monocc float @"HelloWorld.Program:Test1 (single)"(float %arg_x) #0 {
BB0:
  %pow2 = fmul float %arg_x, %arg_x
  ret float %pow2
}

define monocc double @"HelloWorld.Program:Test2 (double)"(double %arg_x) #0 {
BB0:
  %pow2 = fmul double %arg_x, %arg_x
  ret double %pow2
}

define monocc float @"HelloWorld.Program:Test3 (single,single)"(float %arg_x, float %arg_y) #0 {
BB0:
  %.inv = fcmp ogt float %arg_x, %arg_y
  %t26.v = select i1 %.inv, float %arg_y, float %arg_x
  ret float %t26.v
}

define monocc double @"HelloWorld.Program:Test4 (double,double)"(double %arg_x, double %arg_y) #0 {
BB0:
  %.inv = fcmp ogt double %arg_x, %arg_y
  %t21 = select i1 %.inv, double %arg_y, double %arg_x
  ret double %t21
}

define monocc float @"HelloWorld.Program:Test5 (single,single)"(float %arg_x, float %arg_y) #0 {
BB0:
  %.inv = fcmp olt float %arg_x, %arg_y
  %t21.v = select i1 %.inv, float %arg_y, float %arg_x
  ret float %t21.v
}

define monocc float @"HelloWorld.Program:Test5 (single,single)"(float %arg_x, float %arg_y) #0 {
BB0:
  %.inv = fcmp olt float %arg_x, %arg_y
  %t21.v = select i1 %.inv, float %arg_y, float %arg_x
  ret float %t21.v
}
```
CoreCLR: [sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4axAMxNSDIQwDeNGgzMMA2gFkYGABYQAJgElxkgBQ37T12MkB5MT4ILlwWADkIZy5JXi44gHMASgBdU3NmJAYAM0kIbAwGABVVDHJ3XPzChCSGAF4APgYrArsAMRYABQgAd3cENAZSJI1qdLNrWwcXN08pnzdA4NCIqJi4xNTxpnIsxwgOYGli0tJ3fcPjmvqmlvsu3v7B4dHtye8ZvzmP3wCg3hCYUi0Vi8S4yTS1HMOyylQKJ1wGC0FTy8IGOVRhQAnrVGs1Wh0rHEngwca8oeZ3tNft9qYt/oDViCNuCthSzJkGBcjjAERgUOcDjyGOjucccTd8fciVwSWSTOzLF46V9lQs/EsAStgeswRDtpy4YUSoiAKwoqoiwZG0m426tFgtBBykYK6FU9UeNWfP7LIFrUGbSHQzli3kmjBIQWXXmioXiu1SuyO7DO9HyxU0AC+QA==) (it doesn't unroll Math.Pow and doesn't have fast math mode for vmax)

### CODE REVIEW

**Comment by filipnavara** [comment]
## Burgers benchmark

### Before

|    Method |       Mean |   Error |  StdDev |     Median |        Min |        Max |      Gen 0 |      Gen 1 |      Gen 2 | Allocated |
|---------- |-----------:|--------:|--------:|-----------:|-----------:|-----------:|-----------:|-----------:|-----------:|----------:|
| Burgers_0 | 2,665.4 ms | 5.39 ms | 4.21 ms | 2,665.7 ms | 2,654.4 ms | 2,669.3 ms | 46000.0000 | 46000.0000 | 46000.0000 |         - |
| Burgers_1 | 2,251.6 ms | 5.40 ms | 4.79 ms | 2,250.4 ms | 2,245.7 ms | 2,264.1 ms |          - |          - |          - |         - |
| Burgers_2 |   149.9 ms | 0.52 ms | 0.48 ms |   149.7 ms |   149.3 ms |   150.6 ms |          - |          - |          - |         - |
| Burgers_3 | 2,414.9 ms | 5.80 ms | 5.14 ms | 2,412.9 ms | 2,411.1 ms | 2,429.8 ms |          - |          - |          - |         - |

### After

|    Method |       Mean |   Error |  StdDev |     Median |        Min |        Max |      Gen 0 |      Gen 1 |      Gen 2 | Allocated |
|---------- |-----------:|--------:|--------:|-----------:|-----------:|-----------:|-----------:|-----------:|-----------:|----------:|
| Burgers_0 |   537.3 ms | 7.10 ms | 6.64 ms |   537.1 ms |   523.8 ms |   546.2 ms | 45000.0000 | 45000.0000 | 45000.0000 |         - |
| Burgers_1 |   151.6 ms | 1.63 ms | 1.52 ms |   152.2 ms |   149.4 ms |   154.0 ms |          - |          - |          - |         - |
| Burgers_2 |   150.0 ms | 0.45 ms | 0.42 ms |   150.1 ms |   149.3 ms |   150.7 ms |          - |          - |          - |         - |
| Burgers_3 | 2,418.1 ms | 5.25 ms | 4.65 ms | 2,417.6 ms | 2,412.7 ms | 2,430.5 ms |          - |          - |          - |         - |

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/mini/intrinsics.c [hunk 1]
```diff
@@ -163,16 +159,31 @@ llvm_emit_inst_for_method (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSign
...
 			opcode = OP_ABS;
 		}
+		// Max and Min can only be optimized in fast math mode
+		else if (strcmp (cmethod->name, "Max") == 0 && mono_use_fast_math && fsig->params [0]->type == MONO_TYPE_R8) {
+			opcode = OP_RMAX; 
+		} else if (strcmp (cmethod->name, "Min") == 0 && mono_use_fast_math && fsig->params [0]->type == MONO_TYPE_R8) {
+			opcode = OP_RMIN;
+		} else if (strcmp (cmethod->name, "Max") == 0 && mono_use_fast_math && fsig->params [0]->type == MONO_TYPE_R4) {
+			opcode = OP_RMAXF; 
+		} else if (strcmp (cmethod->name, "Min") == 0 && mono_use_fast_math && fsig->params [0]->type == MONO_TYPE_R4) {
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply mono/mini/intrinsics.c]: I guess you mean R8 -> OP_R, R4 -> OP_F

[the remark this answers is not in the document]
Comment (EgorBo) [reply mono/mini/intrinsics.c]: Ouch

[the remark this answers is not in the document]
Comment (EgorBo) [reply mono/mini/intrinsics.c]: renamed, I hope I didn't mess up

**Comment by EgorBo** [comment]
@monojenkins build Windows x64 FullAOT

**Comment by tannergooding** [comment]
`Math.Max` to `vmaxss` is not an equivalent translation. They handle NaN and +/-0 differently, so it is not safe to do unless the user has opted into some kind of "fast math" mode.

**Comment by lewing** [comment]
@tannergooding that is what `--ffast-math` does see the `mono_use_fast_math` checks in the changes.

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**